lib/Conversion/ConvertToArcs/ConvertToArcs.cpp

```diff
@@ -507,29 +511,95 @@ LogicalResult Converter::absorbRegs(HWModuleOp module) {
+                             ConversionPatternRewriter &rewriter,
+                             const TypeConverter &converter) {
+  // Convert the result types.
+  SmallVector<Type> resultTypes;
+  if (failed(converter.convertTypes(op.getResultTypes(), resultTypes)))
+    return failure();
+
+  // Collect the SSA values defined outside but used inside the body region.
+  auto cloneIntoBody = [](Operation *op) {
+    return op->hasTrait<OpTrait::ConstantLike>();
+  };
+  auto operands =
+      mlir::makeRegionIsolatedFromAbove(rewriter, op.getBody(), cloneIntoBody);
+
+  // Create a replacement `arc.execute` op.
+  auto executeOp =
+      ExecuteOp::create(rewriter, op.getLoc(), resultTypes, operands);
+  executeOp.getBody().takeBody(op.getBody());
+  rewriter.replaceOp(op, executeOp.getResults());
+  return success();
+}
+
+/// `llhd.yield` -> `arc.output`
+static LogicalResult convert(llhd::YieldOp op, llhd::YieldOp::Adaptor adaptor,
+                             ConversionPatternRewriter &rewriter) {
+  rewriter.replaceOpWithNewOp<arc::OutputOp>(op, adaptor.getOperands());
+  return success();
+}
+
 //===----------------------------------------------------------------------===//
 // Pass Infrastructure
 //===----------------------------------------------------------------------===//
 
 namespace circt {
-#define GEN_PASS_DEF_CONVERTTOARCS
+#define GEN_PASS_DEF_CONVERTTOARCSPASS
 #include "circt/Conversion/Passes.h.inc"
 } // namespace circt
 
 namespace {
-struct ConvertToArcsPass : public impl::ConvertToArcsBase<ConvertToArcsPass> {
-  using ConvertToArcsBase::ConvertToArcsBase;
-
-  void runOnOperation() override {
-    Converter converter;
-    converter.tapRegisters = tapRegisters;
-    if (failed(converter.run(getOperation())))
-      signalPassFailure();
-  }
+struct ConvertToArcsPass
+    : public circt::impl::ConvertToArcsPassBase<ConvertToArcsPass> {
+  using ConvertToArcsPassBase::ConvertToArcsPassBase;
+  void runOnOperation() override;
 };
 } // namespace
 
-std::unique_ptr<OperationPass<ModuleOp>>
-circt::createConvertToArcsPass(const ConvertToArcsOptions &options) {
-  return std::make_unique<ConvertToArcsPass>(options);
+void ConvertToArcsPass::runOnOperation() {
+  // Setup the type conversion.
+  TypeConverter converter;
+
+  // Define legal types.
+  converter.addConversion([](Type type) -> std::optional<Type> {
+    if (isa<llhd::LLHDDialect>(type.getDialect()))
+      return std::nullopt;
+    return type;
+  });
+
+  // Gather the conversion patterns.
+  ConversionPatternSet patterns(&getContext(), converter);
+  patterns.add<llhd::CombinationalOp>(convert);
+  patterns.add<llhd::YieldOp>(convert);
+
+  // Setup the legal ops. (Sort alphabetically.)
+  ConversionTarget target(getContext());
+  target.addIllegalDialect<llhd::LLHDDialect>();
+  target.markUnknownOpDynamicallyLegal(
+      [](Operation *op) { return !isa<llhd::LLHDDialect>(op->getDialect()); });
```

Comment:
Is it required even when llhd::LHDDialect is registered as an illegal dialect?

Reply:
Yeah I ran into the issue that all operations/dialects not explicitly listed as legal leading to a "failed to legalize" error. I would have thought that the partial dialect conversion would just rewrite ops that were explicitly marked as illegal. But that didn't work for me :thinking:

Reply:
One of the reasons I often just use greedy pattern rewriter or simple walk, as I always have to remember again and mostly I want any pattern that can apply to apply. (dialect conversion still very useful for ones where type conversion needed and I haven't tried the new oneshot one).

Reply:
Yeah same here. I have _no_ mental model of the dialect conversion framework. It feels completely random to me what I have to put into legalization, type conversion, and what gets converted where. But the plain old rewriting framework is annoying when a lot of type conversions are involved. I'm also having trouble finding good documentation for the dialect conversion framework. Seems to be a RTFC kind of deal :grin: